client/galaxy/scripts/mvc/tool/tool-form-composite.js

```diff
@@ -319,6 +319,19 @@ define([ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view'
             }
         },
 
+        /** Refresh the history after job submission while form is shown */
+        _refreshHistory: function() {
+            var self = this;
+            var history = parent.Galaxy && parent.Galaxy.currHistoryPanel && parent.Galaxy.currHistoryPanel;
+            this._refresh_history && clearTimeout( this._refresh_history );
+            if ( history ) {
+                history.refreshContents();
+                if ( history.model.numOfUnfinishedShownContents() === 0 ) {
+                    this._refresh_history = setTimeout( function() { self._refreshHistory() }, 2000 );
+                }
+            }
+        },
+
         /** Build remaining steps */
         _execute: function() {
```

Comment:
Adding an empty job feels like a hackish way of having a proper flag. The flag would be set to True at the start of the WF invocation and set to False after finishing( a problem that immidietally comes to mind is multiple WFs adding datasets to the same history - but this feels solveable).
History polling would continue (and trigger) if the flag on history is True in addition to poll if there are jobs in non-terminal state - this does not sound like a concept-breaking change to me - but @carlfeberhard is the law here.